### PR TITLE
Feature/project

### PR DESCRIPTION
**Pull request by nourelhodaahmed**
This PR introduces key enhancements to the project management flow, covering both core logic implementation and comprehensive unit testing. It affects the following components:
🛠 Implementation Changes
📁 CsvProjectDataSource

    Implemented robust handling for all project-related operations:

        readProjects, writeProject, editProject, deleteProject, and getProject.

    Improved parsing logic and interaction with FileHelper and CsvProjectParser.

    Enhanced error handling to ensure graceful failures on file or parsing exceptions.

    Ensured data integrity when reading, writing, or modifying CSV records.

🧠 ProjectUseCase

    Integrated use case logic for high-level operations on Project entities.

    Applied Clean Architecture principles: use case separation, dependency inversion.

    Designed logic to handle business rules for project creation, editing, and retrieval.

🧪 Test Coverage
CsvProjectDataSourceTest

    Added full unit test coverage:

        Success paths for all CRUD operations.

        Failure scenarios (e.g., file read/write errors, parser issues).

        Parameterized tests to validate behavior on varied outcomes.

- data test cases coverage:

image

- logic test cases coverage:

image

### CODE REVIEW

**Review by ElsayedMagdy122 — APPROVED** [review]

File: src/main/kotlin/data/project/datasource/ProjectColumnIndex.kt
Comment (eslam-mohamed-dev) [inline src/main/kotlin/data/project/datasource/ProjectColumnIndex.kt:1]: you can avoid using object class and use enum and use such ProjectColumnIndex.STATE_ID.ordinal to reach index of each field in its order in the enum

**Review by ElsayedMagdy122 — APPROVED** [review]

**Review by AmrAshraf-git — APPROVED** [review]

File: src/main/kotlin/logic/project/GetAllProjectsUseCase.kt [hunk 1]
```diff
@@ -1,2 +1,13 @@
...
 
+import squad.abudhabi.logic.exceptions.NoProjectsFoundException
+import squad.abudhabi.logic.model.Project
+import squad.abudhabi.logic.repository.ProjectRepository
+
+class GetAllProjectsUseCase(private val projectRepository: ProjectRepository) {
+    operator fun invoke(): List<Project> {
+        val storedProjects = projectRepository.getAllProjects()
+        if (storedProjects.isEmpty()) throw NoProjectsFoundException()
+        return storedProjects
```
Comment (Noor-Serry) [inline src/main/kotlin/logic/project/GetAllProjectsUseCase.kt]: just a small improvement here return projectRepository.getAllProjects()
    .takeIf { it.isNotEmpty() }
    ?: throw NoProjectsFoundException()

File: src/test/kotlin/data/project/datasource/CsvProjectDataSourceTest.kt [hunk 2]
```diff
@@ -0,0 +1,388 @@
...
+    private lateinit var csvProjectParser: CsvProjectParser
+    private lateinit var csvProjectDataSource: CsvProjectDataSource
+
+    @BeforeEach
+    fun setup() {
+        fileHelper = mockk(relaxed = true)
+        csvProjectParser = mockk(relaxed = true)
+        csvProjectDataSource = CsvProjectDataSource(fileHelper, csvProjectParser, "project.csv")
+    }
+
```
Comment (Noor-Serry) [inline src/test/kotlin/data/project/datasource/CsvProjectDataSourceTest.kt:27]: its good practices to start with the happy case scenario

File: src/test/kotlin/data/project/datasource/CsvProjectDataSourceTest.kt [hunk 3]
```diff
@@ -0,0 +1,388 @@
...
+        every { fileHelper.readFile(any()) } returns listOf()
+        every { csvProjectParser.parseProjectToString(any()) } returns "id1,name1,id1-name1"
+        //when
+        csvProjectDataSource.writeProject(project)
+        //then
+        verify(exactly = 1) { fileHelper.writeFile("project.csv", listOf("id1,name1,id1-name1")) }
+    }
+
+    @Test
+    fun `writeProject should return true when readProjects return list`() {
```
Comment (Noor-Serry) [inline src/test/kotlin/data/project/datasource/CsvProjectDataSourceTest.kt]: I think you forget to change function description

File: src/test/kotlin/data/project/datasource/CsvProjectDataSourceTest.kt [hunk 4]
```diff
@@ -0,0 +1,388 @@
...
+        assertThrows<Exception> {
+            csvProjectDataSource.readProjects()
+        }
+    }
+
+    @Test
+    fun `readProjects should returns empty list when read file returns empty list`() {
+        //given
+        every { fileHelper.readFile(any()) } returns listOf()
+        //when & then
```
Comment (Noor-Serry) [inline src/test/kotlin/data/project/datasource/CsvProjectDataSourceTest.kt]: a little improvement here you can separate when and then pattern here like this 
// when 
val result = csvProjectDataSource.readProjects()
// then 
assertThat(result).isEmpty()

File: src/test/kotlin/data/project/repository/ProjectRepositoryImplTest.kt [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+        val res = Project("1", "name1", listOf())
+        every { projectDataSource.writeProject(any()) } throws Exception()
+        //when & then
+        assertThrows<Exception> {
+            projectRepositoryImpl.addProject(res)
+        }
+    }
+
+    @Test
+    fun `addProject best case`(){
```
Comment (Noor-Serry) [inline src/test/kotlin/data/project/repository/ProjectRepositoryImplTest.kt]: I think you should rename it with the pattern we follow it function name should ..... when .....

File: src/test/kotlin/data/project/repository/ProjectRepositoryImplTest.kt [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+        val res = Project("1", "name1", listOf())
+        every { projectDataSource.editProject(any()) } throws Exception()
+        //when & then
+        assertThrows<Exception> {
+            projectRepositoryImpl.editProject(res)
+        }
+    }
+
+    @Test
+    fun `editProject best case`() {
```
Comment (Noor-Serry) [inline src/test/kotlin/data/project/repository/ProjectRepositoryImplTest.kt]: what about editProject should call editProject when there are no error occur

File: src/main/kotlin/data/project/datasource/CsvProjectDataSource.kt [hunk 7]
```diff
@@ -1,2 +1,49 @@
...
+    }
+
+    override fun writeProject(project: Project) {
+        val projects = readProjects()
+        writeProjects(projects + project)
+    }
+
+    override fun editProject(project: Project) {
+        val projects = readProjects().toMutableList()
+        projects.find { it.id == project.id } ?: throw ProjectNotFoundException(project.id)
```
Comment (eslam-mohamed-dev) [inline src/main/kotlin/data/project/datasource/CsvProjectDataSource.kt]: a small thing as we discussed yesterday, we will verify if it is found from use case first then edit don't need to check again

File: src/main/kotlin/data/project/datasource/CsvProjectDataSource.kt [hunk 8]
```diff
@@ -1,2 +1,49 @@
...
+        val newProjects = projects.map { currentProject -> currentProject.isEqualProject(project) }
+        writeProjects(newProjects)
+    }
+
+    override fun deleteProject(projectId: String) {
+        val projects = readProjects().toMutableList()
+        projects.find { it.id == projectId } ?: throw ProjectNotFoundException(projectId)
```
Comment (eslam-mohamed-dev) [inline src/main/kotlin/data/project/datasource/CsvProjectDataSource.kt]: same as edit project

File: src/main/kotlin/logic/exceptions/Exceptions.kt [hunk 9]
```diff
@@ -1,2 +1,23 @@
 package squad.abudhabi.logic.exceptions
 
+open class AppException(msg: String): Exception(msg)
+
+class InvalidStateException(message: String) : AppException(message)
```
Comment (eslam-mohamed-dev) [inline src/main/kotlin/logic/exceptions/Exceptions.kt]: here we agreed to make exceptions without editing their messages

File: src/main/kotlin/logic/project/CreateProjectUseCase.kt [hunk 10]
```diff
@@ -1,2 +1,16 @@
 package squad.abudhabi.logic.project
 
+import squad.abudhabi.logic.model.Project
+import squad.abudhabi.logic.repository.ProjectRepository
+
+class CreateProjectUseCase(
+    private val projectRepository: ProjectRepository
+) {
+    operator fun invoke(projectName: String){
+        val newProject = Project(
```
Comment (eslam-mohamed-dev) [inline src/main/kotlin/logic/project/CreateProjectUseCase.kt:11]: abdallah said on creating project you can pass initial states with it

File: src/main/kotlin/data/project/datasource/CsvProjectDataSource.kt [hunk 11]
```diff
@@ -1,2 +1,49 @@
...
+import squad.abudhabi.data.utils.filehelper.FileHelper
+import squad.abudhabi.logic.exceptions.ProjectNotFoundException
+import squad.abudhabi.logic.model.Project
+
+class CsvProjectDataSource(
+    private val fileHelper: FileHelper,
+    private val csvProjectParser: CsvProjectParser,
+    private val fileName: String
+) : ProjectDataSource {
+    override fun readProjects(): List<Project> {
```
Comment (eslam-mohamed-dev) [inline src/main/kotlin/data/project/datasource/CsvProjectDataSource.kt]: we agreed on naming getAll instead of read to be more readable specially when integrating with api later

File: src/main/kotlin/data/project/datasource/CsvProjectDataSource.kt [hunk 12]
```diff
@@ -1,2 +1,49 @@
...
+        return fileHelper.readFile(fileName)
+            .map(csvProjectParser::parseStringToProject)
+    }
+
+    override fun writeProject(project: Project) {
```
Comment (eslam-mohamed-dev) [inline src/main/kotlin/data/project/datasource/CsvProjectDataSource.kt]: also here may be create instead of write

File: src/main/kotlin/data/project/datasource/CsvProjectDataSource.kt [hunk 13]
```diff
@@ -1,2 +1,45 @@
...
+        writeProjects(newProjects)
+    }
+
+    override fun deleteProject(projectId: String) {
+        val projects = getAllProjects()
+        writeProjects(projects.filter { it.id != projectId })
+    }
+
+    override fun getProject(projectId: String): Project? {
+        return getAllProjects().toMutableList()
```
Comment (eslam-mohamed-dev) [inline src/main/kotlin/data/project/datasource/CsvProjectDataSource.kt]: why .toMutableList() ?

File: src/main/kotlin/data/project/datasource/CsvProjectDataSource.kt [hunk 14]
```diff
@@ -1,2 +1,45 @@
...
+            .find { it.id == projectId }
+    }
+
+    private fun writeProjects(projects: List<Project>) {
+        if (projects.isEmpty()) fileHelper.writeFile(fileName, listOf(",,"))
+        else fileHelper.writeFile(fileName, projects.map(csvProjectParser::parseProjectToString))
+    }
+
+    private fun Project.isEqualProject(project: Project): Project {
+        if (this.id == project.id) return project else return this
```
Comment (eslam-mohamed-dev) [inline src/main/kotlin/data/project/datasource/CsvProjectDataSource.kt:45]: project is a data class so == works fine, did you try project1 == project2 and failed?

File: src/main/kotlin/data/project/datasource/CsvProjectDataSource.kt (hunk 13 above)
Comment (nourelhodaahmed) [reply src/main/kotlin/data/project/datasource/CsvProjectDataSource.kt]: oh yes, i forgot it

File: src/main/kotlin/data/project/datasource/CsvProjectDataSource.kt (hunk 14 above)
Comment (nourelhodaahmed) [reply src/main/kotlin/data/project/datasource/CsvProjectDataSource.kt:45]: i use it in edit project so the only field they are equal is the id as the param is edited project

File: src/main/kotlin/logic/exceptions/Exceptions.kt (hunk 9 above)
Comment (eslam-mohamed-dev) [reply src/main/kotlin/logic/exceptions/Exceptions.kt]: if all exceptions are like:
class CanNotParseProjectException : DataException("can't parse string to project")
with fixed message is clearer to have one message to each exception not a general exception with different messages like InvalidStateException(message: String)

you agree?

File: src/main/kotlin/logic/project/GetAllProjectsUseCase.kt (hunk 1 above)
Comment (eslam-mohamed-dev) [reply src/main/kotlin/logic/project/GetAllProjectsUseCase.kt]: and return in one line would also be great

File: src/test/kotlin/data/project/datasource/CsvProjectDataSourceTest.kt [hunk 15]
```diff
@@ -0,0 +1,370 @@
...
+class CsvProjectDataSourceTest {
+    private lateinit var fileHelper: FileHelper
+    private lateinit var csvProjectParser: CsvProjectParser
+    private lateinit var csvProjectDataSource: CsvProjectDataSource
+
+    @BeforeEach
+    fun setup() {
+        fileHelper = mockk(relaxed = true)
+        csvProjectParser = mockk(relaxed = true)
+        csvProjectDataSource = CsvProjectDataSource(fileHelper, csvProjectParser, "project.csv")
```
Comment (eslam-mohamed-dev) [inline src/test/kotlin/data/project/datasource/CsvProjectDataSourceTest.kt]: i suggest using build/ path to avoid pushing the file to github

File: src/main/kotlin/logic/exceptions/Exceptions.kt (hunk 9 above)
Comment (nourelhodaahmed) [reply src/main/kotlin/logic/exceptions/Exceptions.kt]: yes i agree, don't like passing param to exception.
@ElsayedMagdy122 look here

File: src/test/kotlin/data/project/datasource/CsvProjectDataSourceTest.kt (hunk 15 above)
Comment (nourelhodaahmed) [reply src/test/kotlin/data/project/datasource/CsvProjectDataSourceTest.kt]: i don't get what you mean

File: src/test/kotlin/data/project/datasource/CsvProjectDataSourceTest.kt [hunk 16]
```diff
@@ -0,0 +1,370 @@
...
+        every { fileHelper.readFile(any()) } returns listOf("1,name1,1-state1|2-state2|3-state3")
+        every { csvProjectParser.parseStringToProject(any()) } returns Project("1", "name1", resState)
+        // when
+        val result = csvProjectDataSource.getAllProjects()
+        // then
+        Truth.assertThat(result).isEqualTo(res)
+    }
+
+    @Test
+    fun `createProject should throw Exception when readProjects throw Exception`() {
```
Comment (eslam-mohamed-dev) [inline src/test/kotlin/data/project/datasource/CsvProjectDataSourceTest.kt]: noor serry already added append method to file helper so no need to read the whole list to add to it, just call append and pass your project to it in a list

File: src/main/kotlin/data/project/datasource/CsvProjectDataSource.kt [hunk 17]
```diff
@@ -1,2 +1,44 @@
...
+    private val fileName: String
+) : ProjectDataSource {
+    override fun getAllProjects(): List<Project> {
+        return fileHelper.readFile(fileName)
+            .map(csvProjectParser::parseStringToProject)
+    }
+
+    override fun createProject(project: Project) {
+        val projects = getAllProjects()
+        writeProjects(projects + project)
```
Comment (eslam-mohamed-dev) [inline src/main/kotlin/data/project/datasource/CsvProjectDataSource.kt]: noor serry already added append method to file helper so no need to read the whole list to add to it, just call append and pass your project to it in a list

File: src/test/kotlin/data/project/datasource/CsvProjectDataSourceTest.kt [hunk 18]
```diff
@@ -0,0 +1,370 @@
...
+        every { fileHelper.readFile(any()) } returns listOf("id1,name1,id1-name1")
+        every { csvProjectParser.parseStringToProject(any()) } throws Exception()
+        //when & then
+        assertThrows<Exception> {
+            csvProjectDataSource.editProject(project)
+        }
+    }
+
+    @Test
+    fun `editProject should write nothing when read projects returns empty list`() {
```
Comment (eslam-mohamed-dev) [inline src/test/kotlin/data/project/datasource/CsvProjectDataSourceTest.kt:144]: it won't reach this case as you already check whether the project is already found or not

File: src/test/kotlin/data/project/datasource/CsvProjectDataSourceTest.kt [hunk 19]
```diff
@@ -0,0 +1,370 @@
...
+        every { fileHelper.readFile(any()) } returns listOf("id1,name1,id1-name1")
+        every { csvProjectParser.parseStringToProject(any()) } throws Exception()
+        //when & then
+        assertThrows<Exception> {
+            csvProjectDataSource.deleteProject(projectId)
+        }
+    }
+
+    @Test
+    fun `deleteProject should write nothing when read projects returns empty list`() {
```
Comment (eslam-mohamed-dev) [inline src/test/kotlin/data/project/datasource/CsvProjectDataSourceTest.kt:216]: same as above

File: src/test/kotlin/data/project/datasource/CsvProjectDataSourceTest.kt (hunk 18 above)
Comment (nourelhodaahmed) [reply src/test/kotlin/data/project/datasource/CsvProjectDataSourceTest.kt:144]: just to cover all test scenarios as read project could throw, return list, or return empty list

Comment (eslam-mohamed-dev) [reply src/test/kotlin/data/project/datasource/CsvProjectDataSourceTest.kt:144]: okay if it affects the coverage

**Review by eslam-99 — APPROVED** [review]
